### ISSUE TITLE
Build error in servicedef.h: const member ... can't use default assignment operator

### ISSUE THREAD

**Issue by vphantom** [mbarbon/google-protobuf-dynamic#17]
I don't speak C++ so this looks very cryptic to me, but using g++ 4.9.2 on Debian Jessie with protobuf 3.5.1 freshly compiled and installed, the build errors out.  GCC is unhappy about 4 properties being constant without there being an explicit assignment overload method:

```text
In file included from src/dynamic.cpp:3:0:
src/servicedef.h: In instantiation of ‘void std::vector<_Tp, _Alloc>::_M_insert_aux(std::vector<_Tp, _Alloc>::iterator, const _Tp&) [with _Tp = gpd::MethodDef; _Alloc = std::allocator<gpd::MethodDef>; std::vector<_Tp, _Alloc>::iterator = __gnu_cxx::__normal_iterator<gpd::MethodDef*, std::vector<gpd::MethodDef> >; typename std::_Vector_base<_Tp, _Alloc>::pointer = gpd::MethodDef*]’:
/usr/include/c++/4.9/bits/stl_vector.h:925:28:   required from ‘void std::vector<_Tp, _Alloc>::push_back(const value_type&) [with _Tp = gpd::MethodDef; _Alloc = std::allocator<gpd::MethodDef>; std::vector<_Tp, _Alloc>::value_type = gpd::MethodDef]’
src/dynamic.cpp:212:9:   required from here
src/servicedef.h:16:12: error: non-static const member ‘const string gpd::MethodDef::name’, can’t use default assignment operator
     struct MethodDef {
            ^
src/servicedef.h:16:12: error: non-static const member ‘const string gpd::MethodDef::full_name’, can’t use default assignment operator
src/servicedef.h:16:12: error: non-static const member ‘const bool gpd::MethodDef::client_streaming’, can’t use default assignment operator
src/servicedef.h:16:12: error: non-static const member ‘const bool gpd::MethodDef::server_streaming’, can’t use default assignment operator
In file included from /usr/include/c++/4.9/vector:69:0,
                 from /usr/local/include/google/protobuf/compiler/importer.h:41,
                 from src/dynamic.h:6,
                 from src/dynamic.cpp:1:
/usr/include/c++/4.9/bits/vector.tcc:343:16: note: synthesized method ‘gpd::MethodDef& gpd::MethodDef::operator=(const gpd::MethodDef&)’ first required here 
    *__position = __x_copy;
                ^
error building src/dynamic.o from 'src/dynamic.cpp' at /usr/share/perl/5.20/ExtUtils/CBuilder/Base.pm line 175.
```

Do I have an out of date library somewhere or is this an actual bug?

**Comment by ajiyoshi-vg** [comment]
I met the same probrem and I gave up to build Google::ProtocolBuffers::Dynamic by g++4.x.
My workaround was just using g++6 or g++7.


FYI You can avoid the problem with the `-std=c++11` option. But [when I tried to build Google::ProtocolBuffers::Dynamic with the option](https://github.com/ajiyoshi-vg/google-protobuf-dynamic/commit/d57d90190ebee19f6cfd8d470590aaadb6e524a0), another error occurred.

I reproduced the problem with simplified version of c++ code.

```c++
//main.cc
#include <string>
#include <vector>

struct MethodDef {
        MethodDef(const std::string &_name) :
                name(_name)
        { }

        const std::string name;
};

int main() {
        std::vector<MethodDef> v;
        v.push_back( MethodDef("a") );
}
```

This code uses new C ++ feature, so it will result in a compilation error as it is, then get the same error.

```Dockerfile
# Dockerfile
FROM sl:7

RUN yum install -y gcc-c++
```

```bash
$ docker build -t hoge .
...

# it's ok with `-std=c++11` option
$ docker run -v `pwd`/main.cc:/main.cc -t hoge g++ -std=c++11 main.cc

# got the same build error without option
$ docker run -v `pwd`/main.cc:/main.cc -t hoge g++ main.cc
main.cc: In instantiation of 'void std::vector<_Tp, _Alloc>::_M_insert_aux(std::vector<_Tp, _Alloc>::iterator, const _Tp&) [with _Tp = MethodDef; _Alloc = std::allocator<MethodDef>; std::vector<_Tp, _Alloc>::iterator = __gnu_cxx::__normal_iterator<MethodDef*, std::vector<MethodDef> >; typename std::_Vector_base<_Tp, _Alloc>::pointer = MethodDef*]':
/usr/include/c++/4.8.2/bits/stl_vector.h:913:28:   required from 'void std::vector<_Tp, _Alloc>::push_back(const value_type&) [with _Tp = MethodDef; _Alloc = std::allocator<MethodDef>; std::vector<_Tp, _Alloc>::value_type = MethodDef]'
main.cc:14:33:   required from here
main.cc:4:8: error: non-static const member 'const string MethodDef::name', can't use default assignment operator
 struct MethodDef {
        ^
In file included from /usr/include/c++/4.8.2/vector:69:0,
                 from main.cc:2:
/usr/include/c++/4.8.2/bits/vector.tcc:335:16: note: synthesized method 'MethodDef& MethodDef::operator=(const MethodDef&)' first required here
    *__position = __x_copy;
                ^
```

**Comment by mbarbon** [comment]
Thanks for the follow-up and the additional information, I started looking into this yesterday, but got side-tracked, I will try to dig more today.,

**Comment by mbarbon** [comment]
Fixed with 04e8e8f4a36a51cb4322961816153cf48cc6fee9, I will make a new release later today.

**Comment by vphantom** [comment]
Thank you very much for looking into this @mbarbon!
